### PR TITLE
KAFKA-9801: Still trigger rebalance when static member joins in CompletingRebalance phase

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Fix the direct cause of the observed issue on the client side: when heartbeat getting errors and resetting generation, we only need to set it to UNJOINED when it was not already in REBALANCING; otherwise, the join-group handler would throw the retriable UnjoinedGroupException to force the consumer to re-send join group unnecessarily.

2. Fix the root cause of the issue on the broker side: we should still trigger rebalance when static member joins in CompletingRebalance phase; otherwise the member.ids would be changed when the assignment is received from the leader, hence causing the new member.id's assignment to be empty.

3. Added log4j entries as a by-product of my investigation.

Testing coverage still in progress.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by abbccdda — COMMENTED** [review]
Could we build this PR on top of trunk instead?

**Comment by abbccdda** [comment]
Nvm, should be fine to just rebase on 2.5

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 1]
```diff
@@ -451,8 +451,9 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     return false;
                 }
             } else {
-                resetJoinGroupFuture();
                 final RuntimeException exception = future.exception();
+                log.info("Join group failed with {}", exception.toString());
+                resetJoinGroupFuture();
```
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:456]: What's the reasoning for this move?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 2]
```diff
@@ -892,7 +893,11 @@ protected synchronized String memberId() {
 
     private synchronized void resetGeneration() {
         this.generation = Generation.NO_GENERATION;
-        resetStateAndRejoin();
+        rejoinNeeded = true;
+
+        // only reset the state to un-joined when it is not already in rebalancing
+        if (state != MemberState.REBALANCING)
```
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: Should be equivalent to `state == MemberState.STABLE`?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 3]
```diff
@@ -1025,7 +991,51 @@ class GroupCoordinator(val brokerId: Int,
...
+          protocolName = group.protocolName,
+          // We want to avoid current leader performing trivial assignment while the group
+          // is in stable sync stage, because the new assignment in leader's next sync call
+          // won't be broadcast by a stable group. This could be guaranteed by
+          // always returning the old leader id so that the current leader won't assume itself
+          // as a leader based on the returned message, since the new member.id won't match
+          // returned leader id, therefore no assignment will be performed.
+          leaderId = currentLeader,
+          error = Errors.NONE))
+      case CompletingRebalance =>
```
Comment (abbccdda) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:1032]: We should add a comment about why we would like to trigger rebalance in this case, for example "The rejoining static member will replace the assigned member.id, which could be unknown to the current leader. To avoid having an unknown member assignment returned, we need to trigger another rebalance".

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala [hunk 4]
```diff
@@ -630,30 +630,31 @@ class GroupCoordinatorTest {
...
       sendJoinGroup(groupId, JoinGroupRequest.UNKNOWN_MEMBER_ID, protocolType, protocols, groupInstanceId = followerInstanceId)
     timer.advanceClock(1)
-
-    // Old follower sync callback will return fenced exception while broker replaces the member identity.
     val oldFollowerSyncGroupResult = Await.result(oldFollowerSyncGroupFuture, Duration(1, TimeUnit.MILLISECONDS))
     assertEquals(Errors.FENCED_INSTANCE_ID, oldFollowerSyncGroupResult.error)
 
-    // Duplicate follower will get the same response as old follower.
+    // Duplicate follower joins group with unknown member id will trigger member.id replacement,
+    // and will also trigger a rebalance under CompletingRebalance state
```
Comment (abbccdda) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: Do we have a way to verify the group actually transits CompletingRebalance -> PrepareRebalance -> CompletingRebalance?

**Review by abbccdda — COMMENTED** [review]
Thanks for the PR, the fix itself LGTM, left some comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 2 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: A couple questions on this change:

1. If we are leaving the group through `resetGenerationOnLeaveGroup`, do we still want to continue the rebalance?
2. We use the the `REBALANCING` state to tell whether to return `RebalanceInProgressException` or `CommitFailedException`. If we reset the generation _and_ leave the state in `REBALANCING`, wouldn't that mess up the check?

Mentioned this offline, but if the problem is misinterpreting responses from the wrong generation, I think it would be safer to detect them directly. We can add the generation information to `HeartbeatResponseHandler` for example. If the stored generation does not match the current generation when the response is received, we can ignore it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 5]
```diff
@@ -892,7 +893,11 @@ protected synchronized String memberId() {
 
     private synchronized void resetGeneration() {
         this.generation = Generation.NO_GENERATION;
-        resetStateAndRejoin();
+        rejoinNeeded = true;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Leaving comment here for lack of an alternative. Is it worth checking for the case of an empty assignment? Would the client have been able to recover from this problem by leaving the group and rejoining? I am not sure how much value trying to detect broker versions which are affected by this bug. At least perhaps we can provide a better error message than the BufferUnderflow.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 6]
```diff
@@ -1025,7 +991,51 @@ class GroupCoordinator(val brokerId: Int,
     } else {
       group.removePendingMember(memberId)
     }
-    maybePrepareRebalance(group, s"Adding new member $memberId with group instanceid $groupInstanceId")
+    maybePrepareRebalance(group, s"Adding new member $memberId with group instance id $groupInstanceId")
+  }
+
+  private def updateStaticMemberAndRebalance(group: GroupMetadata,
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:997]: The logic we have in `doSyncGroup` to fill in missing assignments seems dubious. Perhaps there's a compatibility argument to retain it, but it seems more likely to cause problems like the one here than to be useful.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala [hunk 7]
```diff
@@ -630,30 +630,31 @@ class GroupCoordinatorTest {
...
     val duplicateFollowerJoinFuture =
       sendJoinGroup(groupId, JoinGroupRequest.UNKNOWN_MEMBER_ID, protocolType, protocols, groupInstanceId = followerInstanceId)
     timer.advanceClock(1)
-
-    // Old follower sync callback will return fenced exception while broker replaces the member identity.
     val oldFollowerSyncGroupResult = Await.result(oldFollowerSyncGroupFuture, Duration(1, TimeUnit.MILLISECONDS))
     assertEquals(Errors.FENCED_INSTANCE_ID, oldFollowerSyncGroupResult.error)
 
-    // Duplicate follower will get the same response as old follower.
+    // Duplicate follower joins group with unknown member id will trigger member.id replacement,
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: How does this test case reproduce the problem? I thought we would need a sync group from the leader which included the old memberId?

**Review by hachikuji — COMMENTED** [review]
Thanks, left a few comments.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala (hunk 7 above)
Comment (guozhangwang) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: We just need to check that when the group is already in `CompletingRebalance`, a duplicate member (could be follower or leader) sending a join group again (line 638) will still trigger a rebalance.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala (hunk 4 above)
Comment (guozhangwang) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala]: Good call. In line 623 the group is in CompletingRebalance, and then it transits to PrepareRebalance --- let me add the verification of the group state as well.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 6 above)
Comment (guozhangwang) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:997]: Yup I agree, let me file a JIRA for this.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 2 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: There are other states beside STABLE and REBALANCING

Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: 1. Good call, I think the two code path should better not sharing the same callee anyways.
2. I think it would not, since if the member is indeed be participating in a rebalance (i.e. it has already sent a join-group request) and a commit is being failed, then heartbeat failure should not change it to return `CommitFailedException`.

About your general comment, I agree that it's better to let heartbeat request remember the generation when it was sent and compare against it even with this change in place. Let me just do that in this PR.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 1 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:456]: No specific reasoning, I just prefer to change the state after we log the message :)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 5 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Yes that makes sense, let me improve on the error handling.

Regarding recovering from it, that I've thought about and it was a bit tricky --- if broker is on older versions then even if the client handles empty assignment by re-sending a join-group request, it would still be treated as not-triggering a rebalance, then the same empty assignment would still be returned..

**Review by guozhangwang — COMMENTED** [review]
@hachikuji @abbccdda I've addressed your comments and added unit test coverage on client side, also filed https://issues.apache.org/jira/browse/KAFKA-9816. Please take another look at this PR.

**Comment by guozhangwang** [comment]
test this please

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 5 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Yeah, I had the same thought. I'm satisfied with a better error message.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 2 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: Was the first point addressed? As far as I can tell, `resetGenerationOnLeaveGroup` is still going through `resetGeneration`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 8]
```diff
@@ -1008,28 +1013,36 @@ public void handle(LeaveGroupResponse leaveResponse, RequestFuture<Void> future)
...
                         .setMemberId(this.generation.memberId)
                         .setGroupInstanceId(this.rebalanceConfig.groupInstanceId.orElse(null))
                         .setGenerationId(this.generation.generationId));
         return client.send(coordinator, requestBuilder)
-                .compose(new HeartbeatResponseHandler());
+                .compose(new HeartbeatResponseHandler(generation));
     }
 
     private class HeartbeatResponseHandler extends CoordinatorResponseHandler<HeartbeatResponse, Void> {
+        private final Generation sentGeneration;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:1034]: Seems like we'd want to do this for `OffsetCommitResponseHandler` and `SyncGroupResponseHandler` as well? I am thinking this will be a big enough change that we should pull it into a separate PR.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 9]
```diff
@@ -1038,16 +1051,30 @@ public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> futu
                 requestRejoin();
                 future.raise(error);
             } else if (error == Errors.ILLEGAL_GENERATION) {
-                log.info("Attempt to heartbeat failed since generation {} is not current", generation.generationId);
-                resetGenerationOnResponseError(ApiKeys.HEARTBEAT, error);
-                future.raise(error);
+                if (sentGeneration.equals(currentGeneration)) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: This is basically the same thing that we have below. I think we could move it into a generic function in `CoordinatorResponseHandler` so that it could be used by other handlers.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 10]
```diff
@@ -346,6 +346,12 @@ protected void onJoinComplete(int generation,
 
         Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
 
+        // should at least encode the short version
+        if (assignmentBuffer.remaining() < 2)
+            throw new IllegalStateException("Coordinator returned assignment is empty, this is not expected; " +
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Just a nit, but there could be one byte in the message. Maybe we can phrase this message something like "there is insufficient bytes available to read version field (actual size: %d)..."

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 2 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: Regarding the first point, I change the code path to move `this.rejoinNeeded = true;` out of `resetGeneration` and only in `resetGenerationOnResponseError`. But then in trouble-shooting the failed unit test I realized that when heartbeat's poll timer expired it needs to call `leaveGroup` to notify the caller to re-join upon next `poll`. While on the other hand, the only other caller `consumer.close` would close other modules including the coordinator anyways so that right after this call, there will be no heartbeat request or other consumer poll calls, so I think it is actually fine to keep it as is.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 8 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:1034]: SG, I'm reverting this and putting into https://issues.apache.org/jira/browse/KAFKA-9823

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 10 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: ack.

**Comment by mumrah** [comment]
retest this please

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 11]
```diff
@@ -187,11 +187,15 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     assertEquals(1, listener.callsToAssigned)
     assertEquals(0, listener.callsToRevoked)
 
-    Thread.sleep(3500)
+    // as long as we are part of max.poll no rebalance should be triggered
```
Comment (guozhangwang) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: I looked at the flakiness of this test, and propose we just check for exact num.callback

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 2 above)
Comment (abbccdda) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: But we are moving to UNJOINED anyway, so it won't affect the result right?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 12]
```diff
@@ -1427,5 +1440,7 @@ final boolean hasValidMemberId() {
         return generation != Generation.NO_GENERATION && generation.hasMemberId();
     }
 
-
+    final void setNewGeneration(final Generation generation) {
```
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: This seems not used

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java [hunk 13]
```diff
@@ -32,6 +32,7 @@
     private final Timer pollTimer;
 
     private volatile long lastHeartbeatSend = 0L;
+    private RequestFuture<Void> heartbeatFuture = null;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java]: Do we need a reference to the future? A boolean would probably be good enough.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java [hunk 14]
```diff
@@ -56,60 +57,66 @@ public void poll(long now) {
         pollTimer.reset(maxPollIntervalMs);
     }
 
-    public void sentHeartbeat(long now) {
+    boolean hasOngoing() {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java]: nit: maybe `hasInflight`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 2 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: How about `unsubscribe`? That is a case where we send `LeaveGroup`, but we do not intend to rejoin the group.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 15]
```diff
@@ -1317,6 +1328,8 @@ public void onFailure(RuntimeException e) {
                                     }
                                 }
                             });
+
+                            heartbeat.sentHeartbeat(now, heartbeatFuture);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Related to my comment in `Heartbeat`, but I think this reordering is problematic. It is possible that the heartbeat finishes before we reach this point because it is asynchronous. If we use a boolean in `Heartbeat` instead, then we can keep the original ordering.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 12 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Ack

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java (hunk 13 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java]: I have a follow-up PR which leverages on the future which is based on this one --- I think for this PR it does not harm to keep the future object itself although it is only used for boolean checks in unit tests, wdyt?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java (hunk 14 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java]: Sure

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 2 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:903]: My thoughts for `unsubscribe` is that, if consumer still calls `poll` after `unsubscribe` then it needs to either `subscribe` to some other topics or `assign` specifically with topic-partitions. In the first case, the consumer did need to re-join the group, while in the last case, the consumer's coordinator would not call `rejoinNeededOrPending` anyways.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 15 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Cool, that makes sense, I will change it back.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java [hunk 16]
```diff
@@ -465,6 +465,54 @@ public void testSyncGroupRequestWithFencedInstanceIdException() {
...
+
+        coordinator.ensureActiveGroup();
+
+        final AbstractCoordinator.Generation currGen = coordinator.generation();
+
+        // let the heartbeat request to send out a request
+        mockTime.sleep(HEARTBEAT_INTERVAL_MS);
+
+        long startMs = System.currentTimeMillis();
+        while (System.currentTimeMillis() - startMs < 1000 && !coordinator.heartbeat().hasInflight()) {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: nit: we could replace this with `TestUtils.waitForCondition`?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 17]
```diff
@@ -187,11 +187,15 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+    Thread.sleep(500)
+    assertEquals(1, listener.callsToAssigned)
 
-    // we should fall out of the group and need to rebalance
-    awaitRebalance(consumer, listener)
-    assertEquals(2, listener.callsToAssigned)
+    // after we extend longer than max.poll a rebalance should be triggered
+    // NOTE we need to have a relatively much larger value than max.poll to let heartbeat expired for sure
+    Thread.sleep(3000)
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:192]: Can we get rid of this `sleep` since we're blocking below anyway?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 18]
```diff
@@ -187,11 +187,15 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     assertEquals(1, listener.callsToAssigned)
     assertEquals(0, listener.callsToRevoked)
 
-    Thread.sleep(3500)
+    // as long as we are part of max.poll no rebalance should be triggered
+    Thread.sleep(500)
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Can we replace this with `waitUntilTrue` as well?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 17 above)
Comment (guozhangwang) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:192]: Unfortunately we cannot since we need to wait for this amount of time "without polling", and then we should poll and see the rebalance being triggered.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 18 above)
Comment (guozhangwang) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: The num.assignment is already 1 here, and the first sleep is to verify that it did not change; after second thought I think I'll just remove it since without polling there would be no rebalance anyways.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java [hunk 19]
```diff
@@ -494,12 +492,12 @@ public void testHeartbeatUnknownMemberResponseDuringRebalancing() throws Interru
...
         coordinator.requestRejoin();
 
-        startMs = System.currentTimeMillis();
-        while (System.currentTimeMillis() - startMs < 1000 && coordinator.heartbeat().hasInflight()) {
-            Thread.sleep(10);
-            // poll the client until the heartbeat response is received
+        TestUtils.waitForCondition(() -> {
             coordinator.ensureActiveGroup(new MockTime(1L).timer(100L));
-        }
+            return !coordinator.heartbeat().hasInflight();
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:497]: nit: alignment looks a little off

**Review by hachikuji — APPROVED** [review]
LGTM. One minor nit. No need to re-review.

**Comment by ijuma** [comment]
@guozhangwang Checkstyle failed.

**Comment by guozhangwang** [comment]
Cherry-picked to trunk.
